core/kernel/dt.c

```diff
@@ -0,0 +1,220 @@
+#include <mm/core_mmu.h>
+#include <string.h>
+#include <trace.h>
+
+const struct dt_driver *dt_find_driver(const char *compatible)
+{
+	const struct dt_driver *drv = NULL;
+	const struct dt_device_match *id;
+
+	for_each_dt_driver(drv)
+		for (id = drv->match_table; id; id++)
+			if (!strcmp(id->compatible, compatible))
+				break;
+
+	return drv;
+}
+
+const struct dt_driver *__dt_driver_start(void)
+{
+	return &__rodata_dtdrv_start;
+}
+
+const struct dt_driver *__dt_driver_end(void)
+{
+	return &__rodata_dtdrv_end;
+}
+
+int dt_map_dev(const void *fdt, int offs, vaddr_t *base, size_t *size)
+{
+	enum teecore_memtypes mtype;
+	paddr_t pbase;
+	vaddr_t vbase;
+	ssize_t sz;
+	int st;
+
+	assert(cpu_mmu_enabled());
+
+	if (_fdt_get_status(fdt, offs, &st) < 0)
+		return -1;
+	if (st == DT_STATUS_DISABLED)
+		return -1;
+
+	pbase = _fdt_reg_base_address(fdt, offs);
+	if (pbase == (paddr_t)-1)
+		return -1;
+	sz = _fdt_reg_size(fdt, offs);
+	if (sz < 0)
+		return -1;
+
+	if ((st & DT_STATUS_OK_SEC) && !(st & DT_STATUS_OK_NSEC))
+		mtype = MEM_AREA_IO_SEC;
+	else
+		mtype = MEM_AREA_IO_NSEC;
+
+	vbase = (vaddr_t)phys_to_virt(pbase, mtype);
+	if (!vbase) {
+		if (!core_mmu_add_mapping(mtype, pbase, sz)) {
+			EMSG("Failed to map %zu bytes at PA 0x%"PRIxPA,
+			     (size_t)size, pbase);
+			return -1;
+		}
+		vbase = (vaddr_t)phys_to_virt(pbase, mtype);
+		if (!vbase) {
+			EMSG("Failed to get VA for PA 0x%"PRIxPA, pbase);
+			return -1;
+		}
+	}
+
+	*base = vbase;
+	*size = sz;
+	return 0;
+}
+
+/* Read a physical address (n=1 or 2 cells) */
+static paddr_t _fdt_read_paddr(const uint32_t *cell, int n)
+{
+	paddr_t addr;
+
+	if (n < 1 || n > 2)
+		goto bad;
+
+	addr = fdt32_to_cpu(*cell);
+	cell++;
+	if (n == 2) {
+#ifdef ARM32
+		if (addr) {
+			/* High order 32 bits can't be nonzero */
+			goto bad;
+		}
+		addr = fdt32_to_cpu(*cell);
+#else
+		addr = (addr << 32) | fdt32_to_cpu(*cell);
+#endif
+	}
+
+	if (!addr)
+		goto bad;
+
+	return addr;
+bad:
+	return (paddr_t)-1;
+
+}
+
+paddr_t _fdt_reg_base_address(const void *fdt, int offs)
+{
+	const void *reg;
+	int ncells;
+	int len;
+
+	reg = fdt_getprop(fdt, offs, "reg", &len);
+	if (!reg)
+		return (paddr_t)-1;
+
+	ncells = fdt_address_cells(fdt, offs);
+	if (ncells < 0)
+		return (paddr_t)-1;
+
+	return _fdt_read_paddr(reg, ncells);
+}
+
+ssize_t _fdt_reg_size(const void *fdt, int offs)
+{
+	const uint32_t *reg;
+	uint32_t sz;
+	int n;
+	int len;
+
+	reg = (const uint32_t *)fdt_getprop(fdt, offs, "reg", &len);
+	if (!reg)
+		return -1;
+
+	n = fdt_address_cells(fdt, offs);
+	if (n < 1 || n > 2)
+		return -1;
+
+	reg += n;
+
+	n = fdt_size_cells(fdt, offs);
+	if (n < 1 || n > 2)
+		return -1;
+
+	sz = fdt32_to_cpu(*reg);
+	if (n == 2) {
+		if (sz)
+			return -1;
+		reg++;
+		sz = fdt32_to_cpu(*reg);
+	}
+
+	return sz;
+}
+
+static bool is_okay(const char *st, int len)
+{
+	/* Both "ok" and "okay" are valid */
+	if (len > 2)
+		len = 2;
+	return (!strncmp(st, "ok", len));
```

Comment:
> I think it's better to strictly test "ok" and "okay".

Yep.

> Monir: null string returning OK status could also be handle in this function.

That wouldn't help much -- would not work for `secure-status` which defaults to the same value as status when null (i.e., not always "okay").